File: n3/tram.cpp
#include "tram.h"
#include <iostream>
using namespace std;

void Tram::create_tram(const string& tram, const vector<string>& stops) {
    if (tram_to_stops.count(tram)){
        cout << "Трамвай с номером " << " уже существует." << endl; 
    }
    if (stops.size() < 2) {
        cout << "У трамвая должно быть как минимум 2 остановки." << endl;
        return;
    }

    vector<string> cleaned_stops;
    for (const auto& stop : stops) {
        bool exists = false;
        for (const auto& existing : cleaned_stops) {
            if (existing == stop) {
                exists = true;
                break;
            }
        }
        if (!exists) {
            cleaned_stops.push_back(stop);
        }
    }

    if (cleaned_stops.size() < 2) {
        cout << "Были добавлены только 2 одинаковых остановки." << endl;
        return;
    }

    tram_to_stops[tram] = cleaned_stops;
    for (const auto& stop : cleaned_stops) {
        stop_to_trams[stop].insert(tram);
    }
}

void Tram::print_trams_stop(const string& stop, ostream& out) const {
    if (tram_to_stops.empty()) {
        out << "Trams is absent" << endl;
        return;
    }
    if (stop_to_trams.count(stop)) {
        for (const auto& tram : stop_to_trams.at(stop)) {
            out << tram << " ";
        }
        out << endl;
    } else {
        out << "Stop is absent" << endl;
    }
}


void Tram::print_stops_tram(const string& tram, ostream& out) const {
    if (tram_to_stops.empty()) {
        out << "Trams is absent" << endl;
        return;
    }
    if (!tram_to_stops.count(tram)) {
        out << "No tram" << endl;
        return;
    }
    for (const auto& stop : tram_to_stops.at(tram)) {
        out << "Stop " << stop << ": ";
        bool empty = true;
        for (const auto& other : stop_to_trams.at(stop)) {
            if (other != tram) {
                out << other << " ";
                empty = false;
            }
        }
        if (empty) out << "0";
        out << endl;
    }
}
void Tram::print_trams(ostream& out) const {
    if (tram_to_stops.empty()){
        out << "Trams is absent" << endl;
    }
    for (const auto& [tram, stops] : tram_to_stops) {
        out << "TRAM " << tram << ": ";
        for (const auto& stop : stops) {
            out << stop << ", ";
        }
        out << endl;
    } 
}

File: n3/tram.rs
use std::collections::{HashMap, HashSet};
use std::io::{self};

struct TramManager {
    tram_to_stops: HashMap<String, Vec<String>>,
    stop_to_trams: HashMap<String, HashSet<String>>,
}

impl TramManager {
    fn new() -> Self {
        Self {
            tram_to_stops: HashMap::new(),
            stop_to_trams: HashMap::new(),
        }
    }

    fn create_tram(&mut self, tram: String, stops: Vec<String>) {
        if self.tram_to_stops.contains_key(&tram) {
            println!("Ошибка: трамвай с номером {} уже существует.", tram);
            return;
        }
        if self.tram_to_stops.len() >= 1000 {
            println!("Ошибка: достигнуто максимальное количество трамваев (1000).");
            return;
        }
        let mut unique_stops = Vec::new();
        for stop in stops {
            if !unique_stops.contains(&stop) {
                unique_stops.push(stop);
            }
        }
        if unique_stops.len() < 2 {
            println!("Ошибка: должно быть минимум 2 уникальные остановки.");
            return;
        }
        self.tram_to_stops.insert(tram.clone(), unique_stops.clone());
        for stop in unique_stops {
            self.stop_to_trams.entry(stop).or_default().insert(tram.clone());
        }
    }

    fn print_trams_in_stop(&self, stop: &str) {
        if self.tram_to_stops.is_empty() {
            println!("Trams is absent");
            return;
        }
        match self.stop_to_trams.get(stop) {
            Some(trams) => {
                for tram in trams {
                    print!("{} ", tram);
                }
                println!();
            }
            None => println!("Stop is absent"),
        }
    }

    fn print_stops_in_tram(&self, tram: &str) {
        if self.tram_to_stops.is_empty() {
            println!("Trams is absent");
            return;
        }
        match self.tram_to_stops.get(tram) {
            Some(stops) => {
                for stop in stops {
                    print!("Stop {}: ", stop);
                    let mut found = false;
                    if let Some(trams) = self.stop_to_trams.get(stop) {
                        for other in trams {
                            if other != tram {
                                print!("{} ", other);
                                found = true;
                            }
                        }
                    }
                    if !found {
                        print!("0");
                    }
                    println!();
                }
            }
            None => println!("No tram"),
        }
    }

    fn print_all_trams(&self) {
        if self.tram_to_stops.is_empty() {
            println!("Trams is absent");
            return;
        }
        for (tram, stops) in &self.tram_to_stops {
            print!("TRAM {}: ", tram);
            for stop in stops {
                print!("{} ", stop);
            }
            println!();
        }
    }
}

fn main() {
    let mut manager = TramManager::new();
    println!("Введите команды: CREATE_TRAM, TRAMS_IN_STOP, STOPS_IN_TRAM, TRAMS");

    let stdin = io::stdin();
    let mut buffer = String::new();

    while stdin.read_line(&mut buffer).unwrap() > 0 {
        let mut parts = buffer.trim().split_whitespace();
        if let Some(command) = parts.next() {
            match command {
                "CREATE_TRAM" => {
                    if let (Some(tram), Some(count_str)) = (parts.next(), parts.next()) {
                        if let Ok(count) = count_str.parse::<usize>() {
                            let stops: Vec<String> = parts.take(count).map(|s| s.to_string()).collect();
                            if stops.len() == count {
                                manager.create_tram(tram.to_string(), stops);
                            } else {
                                println!("Ошибка: указано {} остановок, но введено меньше.", count);
                            }
                        }
                    }
                }
                "TRAMS_IN_STOP" => {
                    if let Some(stop) = parts.next() {
                        manager.print_trams_in_stop(stop);
                    }
                }
                "STOPS_IN_TRAM" => {
                    if let Some(tram) = parts.next() {
                        manager.print_stops_in_tram(tram);
                    }
                }
                "TRAMS" => manager.print_all_trams(),
                _ => println!("Неизвестная команда: {}", command),
            }
        }
        buffer.clear();
    }
}
